### ISSUE TITLE
accessibility of gui components?

### ISSUE THREAD

**Issue by lilmike** [olilarkin/wdl-ol#53]
Hi,
I stumbled across this library while reading articles across the net on how to get started creating vst/etc plugins, and had a few questions:
1. Are the gui components built into wdl-ol accessible with screen readers? In otherwords, are they simple bitmap graphics, or can they be accessibly represented?
2. If they are not accessible by default, is there a way that they could be made so (if possible, I would like to help with this if that is the case).
3. If they cannot be made accessible for some reason or another, is there a way to use your own gui frameworks/etc with wdl-ol?

As you can see I'm pretty new at all of this, and I mainly ask these questions because there are very few vst synths that are accessible out of the box for blind people, and I'm hoping to develop some of my own to hopefully fill this need.

Thanks,
-Michael.

**Comment by martinfm** [comment]
Hi, 

Having worked with IPlug on [plug-ins specifically designed for visually impaired users ](http://depic.eecs.qmul.ac.uk/apm/), I can share one thing or two about my experience. 

Question 1 
the GUI components of the IPlug framework are not accessible per se. They are self-drawn by the plug-in with the graphic engine of the library (LICE). The oprating systems sees the plug-in as an empy window. Knobs, buttons, sliders etc. are not focusable and cannot be read by the screen reader. 

Question 2
There is in fact a way v.i. people can access your plug-in parameters and the good news is that you need no extra effort to achieve that, as long as you initialize the plug-in parameters, which is something you're meant to do anyway.

The parameters of your plug-in are indeed exposed to the DAW in a nice way by IPlug and some DAW's come with plug-in inspectors that make those parameters accessible from the DAW itself. 
Examples of such DAW's are Cockos Reaper (with OSARA plug-in) and CakeWalk Sonar (with
automation inspector).

So when you type something like :  

`GetParam(kParam)->InitDouble("level", 0, -60, 0, 0.2, "dB")`

you're really telling the DAW that the kParam-th parameter is called "level", it's measured in dB, it ranges from -60 to 0, etc. This information is used by the plug-in inspector to make a sensible representation of the parameter for the screen reader. 

Note that this is not given by default because the underlying representation of the parameter is a float ranging from 0 to 1. So if you don't pack it nicely with the initialization, the plug-in inspector (an in turn the v.i. user) will just get meaningless values in the range from 0 to 1. 

I've seen a number of plug-ins behaving this way and they're not usable by v.i. people. On the contrary sighted people have no problem because they interact with the plug-in through the GUI. The GUI in those cases displays the plug-in values correctly (because the graphics are programed by the author of the plug-in themselves) despite the value being meaningless to the screen reader. 

Question 3
Dunno if you can swap the whole GUI framework that easily. What you can try and do if you're after accessibility is to use native components inside your plug-in. In fact IPlug abstracts over the operating systems but it relies on it for the creation of the window and some of the visual components. For example the ISwitchPopUpControl control creates a native pop up menu. 

IPlug gives you access to the underlying native GUI system. On Windows for instance `HWND 	GetMainWnd ()` returns a handler to the main window of the plug-in. 

So you can try and populate the window with native buttons, faders, selection boxes from the OS and they will be accessible by default. 

I've tried in the past to pursue this route but I couldn't afford it time wise.  
You're gonna have two problems: 

1. you have to manage the whole interaction of your component with the framework yourself 
2. even if you succeed, you might still have the DAW on your way when you try to tab around to focus the components. That is, the DAW will hijack the _tab pressed_ event from your window. There might be ways to work around this though, not sure if it really is an issue. 

hope this helps 
Fiore

**Comment by olilarkin** [comment]
interesting discussion. Reaper excels in this regard because every control is cocoa or win32. Could be achieved in plugins using vwnd (part of wdl) i think
